extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/test/java/io/quarkus/jaxrs/client/reactive/deployment/test/AsyncRestClientFilterTestCase.java

```diff
@@ -0,0 +1,144 @@
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.test.common.http.TestHTTPResource;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class AsyncRestClientFilterTestCase {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(Endpoint.class));
+
+    @TestHTTPResource
+    URL url;
+
+    private Client client;
+
+    @BeforeEach
+    public void before() {
+        client = ClientBuilder.newClient()
+                .register(SyncClientRequestFilter.class)
+                .register(AsyncClientRequestFilter.class)
+                .register(AsyncClientResponseFilter.class);
+    }
+
+    @AfterEach
+    public void after() {
+        client.close();
+    }
+
+    @Test
+    public void test() {
+        Response response = client.target(url.toExternalForm() + "/hello").request().get();
+        assertEquals(201, response.getStatus());
+        MultivaluedMap<String, Object> headers = response.getHeaders();
+        assertEquals("foo", headers.getFirst("sync"));
+        assertEquals("bar", headers.getFirst("async"));
+    }
+
+    public static class Endpoint {
+
+        public void setup(@Observes Router router) {
+            router.route("/hello").handler(new Handler<RoutingContext>() {
+                @Override
+                public void handle(RoutingContext event) {
+                    String syncHeader = event.request().getHeader("sync");
+                    String asyncHeader = event.request().getHeader("async");
+                    event.response()
+                            .putHeader("content-type", "text/plain")
+                            .putHeader("sync", syncHeader)
+                            .putHeader("async", asyncHeader)
+                            .end();
+                }
+            });
+        }
+    }
+
+    @Provider
+    public static class SyncClientRequestFilter implements ClientRequestFilter {
+
+        @Override
+        public void filter(ClientRequestContext context) throws IOException {
+            context.getHeaders().add("sync", "foo");
+        }
+    }
+
+    @Provider
+    public static class AsyncClientRequestFilter implements ResteasyReactiveClientRequestFilter {
+
+        @Override
+        public void filter(ResteasyReactiveClientRequestContext requestContext) {
+            requestContext.suspend();
+            Uni.createFrom().completionStage(CompletableFuture.supplyAsync(() -> "bar"))
+                    .onItem().delayIt().by(Duration.ofSeconds(3))
```

Comment:
We should really come up with a better way of adding delays of 3 seconds to tests, because ultimately they add up. I'm guilty of this too, and I don't know what a better alternative is, and it doesn't necessarily have to be resolved for _this_ PR.

Do you think it would be better to wait not for 3 seconds, but for an event such as "current request handler step ended?" which we can trigger when we exit the handler loop and suspended? Would that have the same sort of semantics we're aiming for with those 3 second sleeps?

Reply:
The reason I wanted to add 3 seconds, is because it exceeds the default timeout of the vert.x event loop and would trigger  a prominent warning.

I am sure we could come up with something more elaborate like you describe :)